File: 2020/rust/aoc/src/day2.rs
fn extract_psk_and_rule(string: &str) -> (usize, usize, char, String) {
    let re = regex::Regex::new(r"(\d+)-(\d+) ([a-z]): (.+)").unwrap();
    let cap = re.captures(string).unwrap();
    (
        cap[1].parse().unwrap(),
        cap[2].parse().unwrap(),
        cap[3].chars().nth(0).unwrap(),
        cap[4].into(),
    )
}

fn part1(data: &[String]) -> usize {
    data.iter()
        .filter(|line| {
            let (min, max, letter, word) = extract_psk_and_rule(line);
            let letter_count = word.chars().filter(|x| *x == letter).count();
            letter_count >= min && letter_count <= max
        })
        .count()
}

fn part2(data: &[String]) -> usize {
    data.iter()
        .filter(|line| {
            let (first, last, letter, word) = extract_psk_and_rule(line);
            (word.chars().nth(first - 1).unwrap() == letter)
                != (word.chars().nth(last - 1).unwrap() == letter)
        })
        .count()
}

#[cfg(test)]
mod tests {
    #[test]
    fn day2_part1() {
        let testdata = [
            "1-3 a: abcde".to_string(),
            "1-3 b: cdefg".to_string(),
            "2-9 c: ccccccccc".to_string(),
        ];
        assert_eq!(2, crate::day2::part1(&testdata));
        assert_eq!(603, crate::day2::part1(&crate::parsers::day2()));
    }

    #[test]
    fn day2_part2() {
        let testdata = [
            "1-3 a: abcde".to_string(),
            "1-3 b: cdefg".to_string(),
            "2-9 c: ccccccccc".to_string(),
        ];
        assert_eq!(1, crate::day2::part2(&testdata));
        assert_eq!(404, crate::day2::part2(&crate::parsers::day2()));
    }
}

File: 2020/fsharp/AOC/Day2.fs
module AOC.Day2

open System
open System.Text.RegularExpressions

let splitInFour (str: string)  = 
    let m = Regex.Match (str, "^(\d+)-(\d+) ([a-z]): (.+)$")
    let cap = Array.tail [| for g in m.Groups -> g.Value |]
    (int cap.[0], int cap.[1], cap.[2].[0], cap.[3])


// 4.399 ms
let part1 data = 
    let isValid min max letter word =
        let getCount lett pword = 
            pword 
            |> Seq.filter (fun x -> x = lett)
            |> Seq.length
        min <= getCount letter word && max >= getCount letter word
    data 
    |> Seq.filter (splitInFour >> 
                    (fun (min, max, letter, word) -> isValid min max letter (Seq.toList word)))
    |> Seq.length

// 2.649 ms
let part2 data =
    let isValid first last letter (word: string) =
        (word.[first-1] = letter) <> (word.[last-1] = letter)
    data 
    |> Seq.filter (splitInFour >> (fun (min, max, letter, word) -> isValid min max letter word))
    |> Seq.length
